### ISSUE TITLE
Can cChunkData be arrays of atomics?

### ISSUE THREAD

**Issue by tigerw** [cuberite/cuberite#2357]
Would it eliminate the need for locked access to chunks' blocks/metas/skylights/blocklights? I'm thinking since the `asyncpathfinder` branch is working, but with just that potential data race, this proposal could simplify matters.

**Comment by worktycho** [comment]
This is  horrible idea, because std::atomic is not a POD class, which gives huge advantages. Also a lot of accesses would still need protection due to race conditions involving multiple blocks.

**Comment by worktycho** [comment]
Also the proposed benefit ignores the serious performance issues in the asyncpathfinder branch. Spinning up a thread to run one step is not acceptable.

**Comment by tigerw** [comment]
About asyncpathfinder:
According to http://stackoverflow.com/questions/14351352/will-asynclaunchasync-in-c11-make-thread-pools-obsolete-for-avoiding-expen Linux thread creation is cheap and Windows uses a thread pool in its implementation.

Forgive me if I misunderstood your meaning, but "one step" in this sense is an entire path calculation. Perhaps the Step() function is badly named.

So in theory, for lots of entities, lots of threads performing pathfinding causes less strain on the tick thread and thus a smoother playing experience than the tick thread doing 10 StepOnce() calls per entity, distributed over 20 ticks. I think, but am not sure, that fitting 10 × 20 × 100 (mobs, say) potentially long running (brief test: 10ms - 600ms) calls plus all the other stuff that happens in a tick into one second could be hard.

Of course, creating lots of threads still isn't ideal. However, that could be solved by optimising cMonster to only make a new cPath when absolutely necessary.

**Comment by worktycho** [comment]
What about os x? It appears at first glance to have an expensive implementation. It is my belief that we need an implementation that guarantees the use of a thread pool. That means we need something other than the standard library.

**Comment by tigerw** [comment]
According to [this](https://xenakios.wordpress.com/2014/09/29/concurrency-in-c-the-cross-platform-way/) OSX has the Grand Central Dispatch library which provides fast concurrency. Example async implementation for there:

```cpp
#include "dispatch/dispatch.h"
template <typename F, typename ...Args>
auto async(F&& f, Args&&... args)
-> std::future<typename std::result_of<F (Args...)>::type>
{
    using result_type = typename std::result_of<F (Args...)>::type;
    using packaged_type = std::packaged_task<result_type ()>;
 
    auto p = new packaged_type(std::bind(std::forward<F>(f), std::forward<Args>(args)...));
    auto result = p->get_future();
    dispatch_async_f(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
                     p, [](void* f_) {
        packaged_type* f = static_cast<packaged_type*>(f_);
        (*f)();
        delete f;
    });
 
    return result;
}
```

**Comment by worktycho** [comment]
Do you want to go through every OS I can name and provide an implementation for each of them? The problem is in the spec.

**Comment by tigerw** [comment]
Maybe slow thread creation on a minority OS is still worse than definite low TPS on a large server though.

**Comment by worktycho** [comment]
And where is a situation where std::async is the only position on the tradeoff? A better implementation for the server at the moment is a dedicated worker thread with work queue. Its a common pattern, we have the infrastructure, and you don't have to deal with spin up costs.

**Comment by tigerw** [comment]
With the amount of bugs there are already and the lack of people fixing them, relying on the standard library seems a more attractive solution than investing effort into making our own thread pool.

Of course, if it's done well, I have nothing to complain about, but random deadlocks or bugs in asynchronous code due to an oversight will be another issue added to a growing list that may or may not be fixed.

**Comment by worktycho** [comment]
I'm not suggesting a thread pool, though we will need one for lua at some point. I'm suggesting something more like the chunksender where we have a single thread that waits on a queue for paths, and just handles paths. Its easy to move over to a pool in the future, and should give better performance than std::async, given nothing else is using it. Of course the other option would be to use a library with a pretested thread pool.

**Comment by tigerw** [comment]
I see. Will you accept the current asyncpathfinder as a stopgap or will you prefer to wait?

**Comment by worktycho** [comment]
I suggest waiting.
